### ISSUE TITLE
Leshan client doesn't retry to initiate bootstrap

### ISSUE THREAD

**Issue by raysad** [eclipse/leshan#701]
Hi,
I was trying to make **leshan demo client** retry initiating bootstrap if it fails. When I looked at the code it's already implemented in **leshan-client-core**  [**RegistrationEngine**](https://github.com/eclipse/leshan/blob/master/leshan-client-core/src/main/java/org/eclipse/leshan/client/servers/RegistrationEngine.java).
``` java
if (dmServers == null || dmServers.isEmpty()) {
    scheduleClientInitiatedBootstrap(BS_RETRY);
    }
```
but it doesn't pass the if statement
``` java 
if (bootstrapFuture == null || bootstrapFuture.isDone() || bootstrapFuture.isCancelled()) {
    ...
}
```
I think the problem is that the previous bootstrap task didn't get terminated.
I solved the problem by adding ``` bootstrapFuture = null ``` when bootstrap fail but i'm sure there is a better way to do it.

**Comment by sbernard31** [comment]
It sounds like a bug ! Thx for reporting this.

I create a PR #702 which should fix this. (Please tell if it works for you)

> I solved the problem by adding bootstrapFuture = null when bootstrap fail but i'm sure there is a better way to do it.

I finally do as you proposed.

I first think to just put the scheduling in another task like this : 
```java
Collection<Server> dmServers = clientInitiatedBootstrap();
if (dmServers == null || dmServers.isEmpty()) {
    schedExecutor.submit(new Runnable() {
        @Override
        public void run() {
            scheduleClientInitiatedBootstrap(BS_RETRY);
        }
    });
}
```
but I think there is no guarantee that the current bootstrap task will finish first ... 

I also think about adding a `boolean force` attribute to scheduleClientInitiatedBootstrap but finally I prefer to go with your proposition with a good comment.

**Comment by sbernard31** [comment]
Just to let you know, 
Reading the Registration engine code to understand this issue, I suspect there is a race condition with `client.triggerRegistrationUpdate()` (see #703)

**Comment by raysad** [comment]
Thank you for your response. And Yes it works for me.

**Comment by sbernard31** [comment]
Thx to confirm it :+1: 
I integrated the fix #702 in master now.
